# Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

1. **Таблица `books`**

In [3]:
query = '''
SELECT 
    *
FROM books
'''

In [4]:
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В базе содержится информация о 1000 книг.

2. **Таблица `authors`**

In [5]:
query = '''
SELECT 
    *
FROM authors
'''

In [6]:
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

Всего 636 авторов.

In [7]:
query = '''
SELECT 
    *
FROM publishers
'''

In [8]:
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Есть информация о 340 издательствах.

4. **Таблица `ratings`**

In [9]:
query = '''
SELECT 
    *
FROM ratings
'''

In [10]:
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В базе содержится информация о 6456 оценках книг.

5. **Таблица `reviews`**

In [11]:
query = '''
SELECT 
    *
FROM reviews
'''

In [12]:
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В базе содержится информация о 2793 обзорах на книги.

**Задания**:
1. Сколько книг вышло после 1 января 2000 года

In [13]:
pd.io.sql.read_sql('''
SELECT COUNT(book_id) AS cnt
FROM books
Where publication_date > '2000-01-01'
''', con = engine)

cnt
0  819

Вывод: в базе данных содержится информация о 819 книгах, дата публикации которых позже 1 января 2000 года.

2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [14]:
pd.io.sql.read_sql('''SELECT books.book_id AS book_id,
                             books.title AS title,
                             review_cnt,
                             average_rating
                   FROM books
                   LEFT JOIN 
                            (SELECT book_id,
                                    COUNT(review_id) AS review_cnt
                             FROM reviews
                             GROUP BY book_id) AS a ON books.book_id = a.book_id
                   LEFT JOIN
                            (SELECT book_id,
                                    AVG(rating) AS average_rating
                             FROM ratings
                             GROUP BY book_id) AS b ON books.book_id = b.book_id
''', con = engine)

book_id                                              title  review_cnt  \
0        652          The Body in the Library (Miss Marple  #3)         2.0   
1        273                                          Galápagos         2.0   
2         51                           A Tree Grows in Brooklyn         5.0   
3        951  Undaunted Courage: The Pioneering First Missio...         2.0   
4        839                                        The Prophet         4.0   
..       ...                                                ...         ...   
995       64                                Alice in Wonderland         4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)         2.0   
997      148                                          Christine         3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)         2.0   
999      828                           The Plot Against America         2.0   

     average_rating  
0          4.500000  
1          4.500000  
2          4.250000  
3          4.000000  
4          4.285714  
..              ...  
995        4.230769  
996        5.000000  
997        3.428571  
998        3.500000  
999        3.000000  

[1000 rows x 4 columns]

Вывод: Вывели нужную информацию по всем книгам.

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [15]:
pd.io.sql.read_sql('''
SELECT publishers.publisher AS name_publisher,
       COUNT(Distinct book_id) AS more_50_pages_books
FROM books
LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE num_pages >50
GROUP BY publishers.publisher
ORDER BY more_50_pages_books DESC
LIMIT 1
''', con = engine)

name_publisher  more_50_pages_books
0  Penguin Books                   42

Вывод: издательство, выпустившее наибольшее число книг, называется Penguin Books. Выпустило 42 книги.

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [16]:
pd.io.sql.read_sql('''
SELECT authors.author,
       AVG(ratings_1.avg_rating) AS avg_rating
FROM books
LEFT JOIN (SELECT book_id,
                  AVG(rating) AS avg_rating,
                  COUNT(rating_id) AS rating_cnt
           FROM ratings
           GROUP BY book_id) AS ratings_1 ON ratings_1.book_id = books.book_id
LEFT JOIN authors ON authors.author_id = books.author_id
WHERE rating_cnt >= 50
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
''', con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

Вывод: Дж. Роулинг имеет самую высокую среднюю оценку. Средняя оценка её книг составляет 4.28 балла.

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [17]:
pd.io.sql.read_sql('''
WITH
ratings_subquery AS (
        SELECT ratings.username,
               COUNT(rating_id) AS ratings_cnt
        FROM ratings
        GROUP BY ratings.username
        HAVING COUNT(rating_id) > 50), 
reviews_subquery AS (
        SELECT reviews.username,
               COUNT(review_id) AS reviews_cnt
        FROM reviews
        GROUP BY reviews.username)
SELECT AVG(reviews_cnt) as avg_reviews_cnt
FROM ratings_subquery
LEFT JOIN reviews_subquery ON reviews_subquery.username = ratings_subquery.username
''', con = engine)

avg_reviews_cnt
0        24.333333

Вывод: среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

**Общий вывод**

В ходе данного проекта была проанализирована база данных сервиса для чтения книг по подписке. В ней содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. В результате анализа получены следующие данные:

* В базе данных содержится информация о 1000 книг, 636 авторах, 340 издательствах, 6456 оценках книг и о 2793 обзорах на книги. 
* В базе данных содержится информация о 819 книгах, дата публикации которых позже 1 января 2000 года.
* Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги.
* Самые высокие оценки у автора Дж. Роулинг. Средняя оценка ее книг составляет 4.28 балла.
* В базе есть информация об активных пользователях - в среднем такие пользователи оставляют 24 обзора.